### ISSUE TITLE
Intersection, Union depends on type variable resolution

### ISSUE THREAD

**Issue by rotu** [microsoft/TypeScript#56906]
### 🔎 Search Terms

generic type variable intersection union order

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Generics

### ⏯ Playground Link

https://www.typescriptlang.org/dev/bug-workbench/?#code/C4TwDgpgBAqgdgSwPZwDwEEA0UBCA+KAXinSgB9cAoUSWRFdOEVAFQOPmTRewEMm81cNHgBrOEgDucTilbs6XVAFc44qXGxtKlAPS6owABYQAztFNGkygDYATKLxs2oAI2jGLvALYfhb5WBHACdoUNNIAGNgBAA3CBsQAC5HJj40qFV1aSghWhh0AAYiRTksiWl0kEF9KDqAPQB+PJF0AEYS2ThGZnKNGoMG5ppWgCZOtQqZejR+ar1BqCadEagASThgCGDzaIxsfBLSADIqVY2tnYhonvkSi+3d4FYqwXPNx+vgADkIeOC7sQHlc9nA-tstIIFoYTOYoJZrPZHM43B4TPCfH5aK5AiEwmYojF4okUmD-lVsGTti0SN9ikCPiDnnNKeDggM6kthv50N8OgzLk9blT2dChjTeeMBZ9or9-qg5hzxUA

### 💻 Code

```ts repro
type Union<A, B> = A | B
type UnionAny<T> = Union<T, any>
type UnknownUnion<T> = Union<unknown, T>

// these should all be the same type but are respectively: any, any, unknown 
type UA0 = Union<unknown, any>
//   ^?
type UA1 = UnionAny<unknown>
//   ^?
type UA2 = UnknownUnion<any>
//   ^?

type Intersect<A, B> = A & B
type IntersectAny<T> = Intersect<T, any>
type IntersectNever<T> = Intersect<never, T>

// these should all be the same type but are respectively: never, any, never
type AN0 = Intersect<any, never>
//   ^?
type AN1 = IntersectAny<never>
//   ^?
type AN2 = IntersectNever<any>
//   ^?

```

### 🙁 Actual behavior

The types `UA0`, `UA1` are resolved as `any` and `UA2` is resolved as `unknown`. These all map down to the same union of `unknown` and `any`, so should all be the same type.

The types `AN0`, `AN2` are resolved as `never` and `AN1` resolved as `any`. These all map down to be the intersection of `any` and `never`, so should all be the same type.

### 🙂 Expected behavior

I expect `UA0`, `UA1` and `UA2` to resolve as the same type (probably `any`)
I expect `AN0`, `AN1` and `AN2` to resolve as the same type (probably `never`)

**Comment by rotu** [comment]
The issue may be that `unknown` is assumed to be an absorbing element for `|` and `any` an absorbing element for `&`. I think it's reasonable that a type expression containing type parameters should depend only on the *eventual value* of those type parameters.

**Comment by fatcerberus** [comment]
Generally speaking, both union and intersection type operators are commutative (with the exception of function intersections, which are treated as overloads and are order-dependent), and any union/intersection containing `any` should always reduce to `any`.

**Comment by rotu** [comment]
@fatcerberus gotcha.

> any union/intersection containing any should always reduce to any.

I think it would be logically consistent for either:

- `any | unknown` and `any & never` are both `any`, since `any` represents degeneracy of the type system.
- `T | unknown` is `unknown`, `T & never` is `never`, since `unknown` and `never` bound the type hierarchy and `any` contains no information about the qualified value.

Why should the former be overriding? I was unable to find docs that make this clear, but it feels like an important design choice.

**Comment by fatcerberus** [comment]
> It only distributes when the left-hand side is a type parameter.

Specifically, a *naked* type parameter.  Stuff like `T & U extends ...` or whatever is also not distributive.  Which is why you can write `[T] extends [U] ? ...` to prevent distribution.

**Comment by rotu** [comment]
> > It only distributes when the left-hand side is a type parameter.
> 
> Specifically, a _naked_ type parameter. Stuff like `(T & U) extends ...` or whatever is also not distributive.

Oy. You’re right. And I can’t figure out when that’s a bug or a feature.

For instance `T|T` behaves like the naked type parameter `T` but `T|U` does *not*, even when `T` and `U` are instantiated with the same type!

There’s also the curious case that *even when the LHS does not contain a type parameter*, when the LHS is `any` but the RHS is not, the conditional evaluates to the union of both branches.

**Comment by fatcerberus** [comment]
> There’s also the curious case that even when the LHS does not contain a type parameter, when the LHS is any but the RHS is not, the conditional evaluates to the union of both branches.

Yeah, that’s a separate behavior from distribution and is intentional.

**Comment by rotu** [comment]
> > There’s also the curious case that even when the LHS does not contain a type parameter, when the LHS is any but the RHS is not, the conditional evaluates to the union of both branches.
> 
> Yeah, that’s a separate behavior from distribution and is intentional.

Yes, it’s intentional. The reason for it (`any` is an upper bound of `union`, so we should infer the most general type possible) seems to imply that `unknown extends T ? A : B` should evaluate to `A | B`.
